lib/sycamore/sycamore/writers/pinecone_writer.py

```diff
@@ -0,0 +1,175 @@
+
+        pc = Pinecone(api_key=self._api_key)
+        try:
+            pc.describe_index(self._index_name)
+        except NotFoundException:
+            if self._dimensions is None:
+                raise ValueError(f"dimensions must be specified in order to create a new index {self._index_name}")
+            if self._index_spec is None:
+                raise ValueError(f"index spec must be supplied in order to create a new index {self._index_name}")
+            pc.create_index(
+                name=self._index_name,
+                dimension=self._dimensions,
+                spec=self._index_spec,
+                metric=self._distance_metric,
+                timeout=None,
+            )
+
+    def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+        from pinecone.grpc import PineconeGRPC
+
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
+
+        pc = PineconeGRPC(api_key=self._api_key)
+        index = pc.Index(self._index_name)
+
+        objects = self._extract_pinecone_objects(block)
+        obj_it = iter(objects)
+        batch = list(itertools.islice(obj_it, self._batch_size))
+        async_results = []
+        while len(batch) > 0:
+            async_results.append(index.upsert(vectors=batch, namespace=self._namespace, async_req=True))
+            batch = list(itertools.islice(obj_it, self._batch_size))
+        for r in async_results:
+            r.result()
+
+    @staticmethod
+    def _extract_pinecone_objects(block: Block) -> Iterable[Vector]:
+        def _flatten_metadata(data: Union[dict, list, tuple], prefix="") -> Iterable[Tuple[Any, Any]]:
+            iterator = []  # type: ignore
+            if isinstance(data, dict):
+                iterator = data.items()  # type: ignore
+            if isinstance(data, (list, tuple)):
+                iterator = enumerate(data)  # type: ignore
+            items = []
+            for k, v in iterator:
+                if isinstance(v, (dict, list, tuple)):
+                    if isinstance(v, (list, tuple)) and all(isinstance(innerv, str) for innerv in v):
+                        items.append(((str(k) if len(prefix) == 0 else f"{prefix}.{k}"), v))
+                    else:
+                        inner_values = _flatten_metadata(v, prefix=(str(k) if len(prefix) == 0 else f"{prefix}.{k}"))
+                        items.extend([(innerk, innerv) for innerk, innerv in inner_values])
+                elif v is not None:
+                    items.append(((str(k) if len(prefix) == 0 else f"{prefix}.{k}"), v))
+            return items
+
+        def _metadata_special_cases(data: dict) -> dict:
+            binary = data.get("binary_representation", None)
+            if binary:
+                b64binary = b64encode(binary)
+                strbinary = b64binary.decode("UTF-8")
+                data["binary_representation"] = strbinary
```

Comment:
Is this the right default? It seems like the user is very rarely going to actually want to write the binary_representation out to the vector store. Maybe we default to skipping it and have a property to enable?

Reply:
Are people gonna want to split document components across several dbs? e.g. vectors to pinecone, metadata to postgres, binary to s3?
Not worried about it now just curious if this is a common etl thing.

Reply:
It looks to me like the other two writers are going to drop the binary representation, so I'd do the same here.